### PR TITLE
DPE-102 Add Charm Library 

### PR DESCRIPTION
**Pull request by WRFitch**
## Proposal
This PR implements most of the functionality available in the machine charm, apart from the systemd service management. Multiple pgbouncer instances will be implemented using pebble in a later PR. 

## Context
- No need to review the pgb.py charm library, it's being imported from [the machine charm](https://github.com/canonical/pgbouncer-operator/blob/main/lib/charms/pgbouncer_operator/v0/pgb.py) (unless you really want to).
- This PR must be merged after [pgbouncer container is updated to v0.1.1](https://github.com/canonical/pgbouncer-container/pull/3).

## Release Notes
- Imported pgbouncer library and reworked charm to make use of it
- Removed unnecessary restart-pgbouncer action 
- Updated config and config management to match new functionality
  - this includes implementing max_db_connections derivatives based on available CPU cores. 
- Moved pgbouncer directory to /var/lib/postgresql/pgbouncer 
- implemented pgbouncer reloading when config changes. 

## Testing
- Unit tests have been reworked to fit the new implementation
- Integration tests haven't changed
- Charm has been tested manually. 
- CI has been updated to use new charm container

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -105,228 +110,145 @@ def _on_pgbouncer_pebble_ready(self, event: PebbleReadyEvent) -> None:
...
+        self.unit.status = MaintenanceStatus("Reloading Pgbouncer")
+        logger.info("reloading pgbouncer application")
 
-    def _generate_password(self) -> str:
-        """Generates a secure password.
+        pgb_container = self.unit.get_container(PGB)
+        if not pgb_container.can_connect():
+            self.unit.status = WaitingStatus(
+                "unable to restart pgbouncer - container not connected."
+            )
```
Comment (marceloneppel) [inline src/charm.py:219]: This check can be removed as it's already being checked before the two calls to this method.

File: src/charm.py [hunk 2]
```diff
@@ -105,228 +110,145 @@ def _on_pgbouncer_pebble_ready(self, event: PebbleReadyEvent) -> None:
...
-        Pgbouncer will not apply configuration changes without reloading, so this must be called
-        after each time config files are changed.
-        TODO implement function stub
+        Raises:
+            FileNotFoundError: if there is no file at the given path.
         """
-        logger.info("reloading pgbouncer application")
-        pass
+        pgb_container = self.unit.get_container(PGB)
+        inaccessible = f"pgbouncer container not accessible, cannot find {filepath}"
```
Comment (marceloneppel) [inline src/charm.py:242]: As this variable is only relevant to the inner scope (`if`), it can be moved to inside there.

File: src/charm.py [hunk 3]
```diff
@@ -105,228 +110,145 @@ def _on_pgbouncer_pebble_ready(self, event: PebbleReadyEvent) -> None:
...
-            the `users` dictionary, with usernames from charm config and corresponding generated
-            passwords appended.
-        """
-        for admin_user in self.config["pgb_admin_users"].split(","):
-            if admin_user not in users or users[admin_user] is None:
-                users[admin_user] = self._hash(self._generate_password())
-        return users
-
-    def _get_userlist_from_container(self) -> Dict[str, str]:
+    def _read_userlist(self) -> Dict[str, str]:
```
Comment (marceloneppel) [inline src/charm.py:191]: This function seems not to be used (it's used only in the unit tests).

File: src/charm.py [hunk 4]
```diff
@@ -105,228 +110,145 @@ def _on_pgbouncer_pebble_ready(self, event: PebbleReadyEvent) -> None:
...
+            return
 
         pgb_container.push(
             USERLIST_PATH,
-            userlist,
-            user=self._pgbouncer_user,
-            permissions=0o600,
+            pgb.generate_userlist(userlist),
+            user=PG_USER,
+            permissions=0o777,
```
Comment (marceloneppel) [inline src/charm.py:183]: Is `777` really needed? PgBouncer by default creates this file with `600` when installed through apt.

File: src/charm.py [hunk 5]
```diff
@@ -105,228 +110,145 @@ def _on_pgbouncer_pebble_ready(self, event: PebbleReadyEvent) -> None:
...
+            return
 
         pgb_container.push(
             INI_PATH,
-            pgbouncer_ini,
-            user=self._pgbouncer_user,
-            permissions=0o600,
+            config.render(),
+            user=PG_USER,
+            permissions=0o777,
```
Comment (marceloneppel) [inline src/charm.py:142]: Is 777 really needed? PgBouncer by default creates this file with 600 when installed through apt.

**Review by marceloneppel — APPROVED** [review]
LGTM. Some minor details in the comments.

File: src/charm.py (hunk 5 above)
Comment (WRFitch) [reply src/charm.py]: You're right, this has been reverted back to 600

File: src/charm.py (hunk 4 above)
Comment (WRFitch) [reply src/charm.py]: You're right, this has been reverted back to 600

File: tests/unit/test_charm.py [hunk 6]
```diff
@@ -4,182 +4,159 @@
...
         userlist = pgb_container.pull(USERLIST_PATH).read()
         self.assertEqual('"test-user" "pw"', userlist)
 
-        self.harness.charm._push_userlist(users=None, reload_pgbouncer=True)
-
-    def test_get_userlist_from_container(self):
-        self.harness.update_config({"pgb_admin_users": "test-admin"})
-        self.harness.charm._push_container_config(users={"test-admin": "pass"})
-        users = self.harness.charm._get_userlist_from_container()
+    def test_read_userlist(self):
```
[the remark this answers is not in the document]
Comment (WRFitch) [reply tests/unit/test_charm.py]: removed

File: src/charm.py (hunk 3 above)
Comment (WRFitch) [reply src/charm.py]: removed

File: src/charm.py (hunk 1 above)
Comment (WRFitch) [reply src/charm.py]: removed

File: tox.ini [hunk 7]
```diff
@@ -59,7 +59,7 @@ deps =
     -r{toxinidir}/requirements.txt
 commands =
     coverage run --source={[vars]src_path} \
-        -m pytest --ignore={[vars]tst_path}integration -v --tb native -s {posargs}
+        -m pytest --ignore={[vars]tst_path}integration -v --tb native -s {posargs} --durations=0
```
[the remark this answers is not in the document]
Comment (WRFitch) [reply tox.ini:62]: It's not, I was just interested to see how long the unit tests took

File: src/charm.py [hunk 8]
```diff
@@ -179,60 +147,39 @@ def _push_pgbouncer_ini(self, users: Dict[str, str] = None, reload_pgbouncer=Fal
...
-        try:
-            # Check that we're not updating this file unnecessarily
-            if pgb_container.pull(USERLIST_PATH).read() == userlist:
-                logger.info("updated userlist does not modify existing pgbouncer userlist")
-                return
-        except FileNotFoundError:
-            # there is no existing userlist.txt file, so carry on and add one.
-            pass
+        pgb_container = self.unit.get_container(PGB)
+        if not pgb_container.can_connect():
```
[the remark this answers is not in the document]
Comment (WRFitch) [reply src/charm.py:173]: I've replaced most of the status setting with error handlers, and defined the container as an object field in the constructor - this way we don't have to check the container is connected constantly, but we also aren't suggesting that any container can be passed into these methods. I also documented where status setting is used in the docstring for `_restart_pgbouncer`

**Review by paulomach — APPROVED** [review]
Nice!

**Pull request merged**